### PR TITLE
[FIX] mail: avoid flicker when uploading multiple attachement

### PR DESCRIPTION
**Pull request by phenix-factory**
task-2282327

### CODE REVIEW

File: addons/mail/static/src/models/thread/thread.js [hunk 1]
```diff
@@ -788,7 +788,14 @@ function factory(dependencies) {
...
             const allAttachments = [...new Set(this.originThreadAttachments.concat(this.attachments))]
-                .sort((a1, a2) => a1.id < a2.id ? 1 : -1);
+                  .sort((a1, a2) => {
+                      // Handle the temporary attachments that should be at the
+                      // start of the list. Temporary attachments have a negative id.
+                      if (a1.id > 0 && a2.id < 0) {
+                          return 1;
+                      }
+                      return a1.id < a2.id ? 1 : -1;
+                  });
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:798]: I'll test to better understand what is going on, but I can already tell the indent here should be 4 not 6, and also the PR is missing a test assertion this new order.

File: addons/mail/static/src/models/thread/thread.js [hunk 2]
```diff
@@ -788,7 +788,14 @@ function factory(dependencies) {
...
         _computeAllAttachments() {
             const allAttachments = [...new Set(this.originThreadAttachments.concat(this.attachments))]
-                .sort((a1, a2) => a1.id < a2.id ? 1 : -1);
+                  .sort((a1, a2) => {
+                      // Handle the temporary attachments that should be at the
+                      // start of the list. Temporary attachments have a negative id.
+                      if (a1.id > 0 && a2.id < 0) {
+                          return 1;
+                      }
+                      return a1.id < a2.id ? 1 : -1;
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:797]: Looks much better after this PR.

Just maybe the condition here should be inverted when both ids are below 0?
If the newest attachment should be displayed first, the one with -2 should be before the one with -1, just the inverse of position numbers.

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:797]: So you would have to handle 4 cases: both < 0, both > 0, a < 0 and b > 0, a > 0 and b < 0

File: addons/mail/static/src/models/thread/thread.js [hunk 3]
```diff
@@ -876,7 +876,20 @@ function factory(dependencies) {
...
-                .sort((a1, a2) => a1.id < a2.id ? 1 : -1);
+                .sort((a1, a2) => {
+                    // Handle the temporary attachments that should be at the
+                    // start of the list. Temporary attachments have a negative id.
+                    if (a1.id > 0 && a2.id < 0) {
+                        return 1;
+                    } else if (a1.id < 0 && a2.id < 0) {
+                        return -1;
+                    } else if (a1.id > 0 && a2.id > 0) {
+                        return 1;
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:887]: Should be the same formula as before in this case `a1.id < a2.id ? 1 : -1`

File: addons/mail/static/src/models/thread/thread.js [hunk 4]
```diff
@@ -876,7 +876,20 @@ function factory(dependencies) {
...
         _computeAllAttachments() {
             const allAttachments = [...new Set(this.originThreadAttachments.concat(this.attachments))]
-                .sort((a1, a2) => a1.id < a2.id ? 1 : -1);
+                .sort((a1, a2) => {
+                    // Handle the temporary attachments that should be at the
+                    // start of the list. Temporary attachments have a negative id.
+                    if (a1.id > 0 && a2.id < 0) {
+                        return 1;
+                    } else if (a1.id < 0 && a2.id < 0) {
+                        return -1;
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:885]: should be reversed when both are negative `a1.id > a2.id ? 1 : -1`

File: addons/mail/static/src/models/thread/thread.js [hunk 5]
```diff
@@ -876,7 +876,20 @@ function factory(dependencies) {
...
+                    } else if (a1.id > 0 && a2.id > 0) {
+                        return 1;
+                    } else if (a1.id < 0 && a2.id > 0) {
+                        return -1;
+                    }
+                    return 1;
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:891]: this can never happen

File: addons/mail/static/src/models/thread/thread.js [hunk 6]
```diff
@@ -877,7 +877,20 @@ function factory(dependencies) {
          */
         _computeAllAttachments() {
             const allAttachments = [...new Set(this.originThreadAttachments.concat(this.attachments))]
-                .sort((a1, a2) => a1.id < a2.id ? 1 : -1);
+                .sort((a1, a2) => {
+                    // Handle the temporary attachments that should be at the
+                    // start of the list. Temporary attachments have a negative id.
```
Comment (alexkuhn) [inline addons/mail/static/src/models/thread/thread.js:882]:
```js
// uploading attachment before uploaded attachments
```

File: addons/mail/static/src/models/thread/thread.js [hunk 7]
```diff
@@ -877,7 +877,20 @@ function factory(dependencies) {
...
+                    if (!a1.isTemporary && a2.isTemporary) {
+                        return 1;
+                    }
+                    if (a1.isTemporary && !a2.isTemporary) {
+                        return -1;
+                    }
+                    // if none of the attachment are temporary or both are
+                    // temporary we sort them based on the absolute id. The
+                    // sort condition is the same in both case.
```
Comment (alexkuhn) [inline addons/mail/static/src/models/thread/thread.js:891]:
```js
// most recent to oldest attachments
```

**Review by alexkuhn — COMMENTED** [review]
Suggestion for comments.
The tips for short and very nice comment is to be at conceptual level.
Don't tell what the code does, or what it is expected to do.

(don't focus too much on the **don't**, actually think in terms of the _concepts_, and you should find very good comment because not constrained by an actual process... yeah it's quite abstract, it feels rather natural for me but it's surprisingly hard to explain...)

**Review by alexkuhn — APPROVED** [review]
@robodoo r+

File: addons/mail/static/src/models/thread/thread.js (hunk 1 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js]: This comment was not done btw, indent still wrong and missing a test, will make new task.

**Pull request closed without merging**